### PR TITLE
Script for load tests

### PR DESCRIPTION
**Pull request by arieljbirnbaum**
Implemented in preparation for the EVM campaign.
Adding to the repo as decided in #1995.
Duplicates code from the existing test suites to allow for modification; should eventually be refactored to have a single source of truth. Consider along #1153, #1878.

## TODO

- [x] Exercise activate/deactivate identity steps
- [x] Run on CI **with a very small number of threads/iterations** to make sure the test code continues to work as intended.

### CODE REVIEW

File: .nvmrc
Comment (arieljbirnbaum) [inline .nvmrc:1]: drive-by

File: tee-worker/ts-tests/integration-tests/common/type-definitions.ts
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/integration-tests/common/type-definitions.ts:1]: drive-by

File: tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts [hunk 1]
```diff
@@ -192,7 +192,7 @@ export default {
                 A1: "Null",
                 A2: "Bytes",
                 A3: "(Bytes,Bytes,Bytes)",
-                A4: "u128",
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/parachain-api/prepare-build/interfaces/identity/definitions.ts]: I don't think this should be here; probably a leftover after rebasing from `staging`.

File: tee-worker/ts-tests/parachain-api/src/index.ts [hunk 2]
```diff
@@ -19,7 +19,9 @@ export type {
     WorkerRpcReturnValue,
     TrustedCallSigned,
     Getter,
+    PublicGetter,
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/parachain-api/src/index.ts:26]: Added export

File: tee-worker/ts-tests/stress-runner [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#! /usr/bin/bash
+
+mkdir -p stress-output
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/stress-runner:3]: TODO: take the number of processes and output dir as a parameter; set correct working directory.

File: tee-worker/ts-tests/stress/bin/stress.ts [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+function randomEvmWallet(): ethers.Wallet {
+    return ethers.Wallet.createRandom();
+}
+function randomWallet(): Wallet {
+    return Math.random() > 0.5
+        ? { type: "substrate", keyringPair: randomSubstrateWallet() }
+        : { type: "evm", wallet: randomEvmWallet() };
+}
+
+const apiContextManager = (
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/stress/bin/stress.ts]: [`using`](https://github.com/tc39/proposal-explicit-resource-management) at home.

The test needs to establish several connections, create new ones on the fly in case of failure, and avoid leaking resources. These `ContextManager` objects help with that.

File: tee-worker/ts-tests/stress/config.json [hunk 5]
```diff
@@ -0,0 +1,11 @@
+{
+	"timerEpoch": 1692230400000,
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/stress/config.json]: TODO: not sure how useful this actually is in practice?

File: tee-worker/ts-tests/stress/config.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+	"connections": 20,
+	"iterations": 10,
+	"substrateEndpoint": "ws://tee-staging.litentry.io:9944",
+	"workerEndpoint": "wss://tee-staging.litentry.io:2000",
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/stress/config.json]: TODO: need a way to run this against local worker on CI. Config should prob. be a command line argument?

File: tee-worker/docker/lit-test-stress-script.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+            context: ..
+            dockerfile: build.Dockerfile
+            target: deployed-client
+        depends_on:
+            litentry-node:
+                condition: service_healthy
+            litentry-worker-1:
+                condition: service_healthy
+            litentry-worker-2:
+                condition: service_healthy
```
Comment (arieljbirnbaum) [inline tee-worker/docker/lit-test-stress-script.yml]: TODO: should probably drop this one? /cc @Kailai-Wang @felixfaisal

File: tee-worker/ts-tests/stress/src/context-manager.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    constructor(enter: ContextFactory<Context>) {
+        this.enter = enter;
+    }
+
+    async do<Result>(task: (context: Context) => Promise<Result>) {
+        const { context, exit } = await this.enter();
+        try {
+            return await task(context);
+        } finally {
+            exit();
```
Comment (arieljbirnbaum) [inline tee-worker/ts-tests/stress/src/context-manager.ts]:
```suggestion
            await exit();
```

Comment (arieljbirnbaum) [reply tee-worker/ts-tests/stress/src/context-manager.ts]: (letting the current CI run finish before I commit this)

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -586,6 +586,7 @@ jobs:
           - test_name: lit-di-evm-identity-test
           - test_name: lit-parentchain-nonce
           - test_name: lit-ii-batch-test
+          - test_name: lit-test-stress-script
```
Comment (Kailai-Wang) [inline .github/workflows/ci.yml:590]: Are we planning to trigger it for each CI run?

File: tee-worker/docker/lit-test-stress-script.yml (hunk 7 above)
Comment (Kailai-Wang) [reply tee-worker/docker/lit-test-stress-script.yml]: yea if you are only sending requests to worker-1 you don't have to wait for worker-2 (it's still launched though)

File: tee-worker/ts-tests/stress/test/test.ts [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+
+function getConfig(): Config {
+    return {
+        connections: 3,
+        iterations: 2,
+        workerEndpoint: process.env.WORKER_ENDPOINT!, // @fixme evil assertion; centralize env access
+        substrateEndpoint: process.env.NODE_ENDPOINT!, // @fixme evil assertion; centralize env access
+    };
+}
+
```
Comment (0xverin) [inline tee-worker/ts-tests/stress/test/test.ts:31]: Can we add a comment here to describe the entire workflow? This could be more user-friendly for others who want to understand the stress test. :P

Comment (0xverin) [reply tee-worker/ts-tests/stress/test/test.ts:31]: To be honest, I need quite some time to digest all of this.

**Review by 0xverin — COMMENTED** [review]
Thanks!But whether to add it to the CI might need some consideration. The benefit of integrating it into CI is to ensure that this stress test remains usable, but it could increase maintenance efforts. I'm concerned that it might add complexity for other developers.

File: tee-worker/ts-tests/stress/test/test.ts (hunk 10 above)
Comment (arieljbirnbaum) [reply tee-worker/ts-tests/stress/test/test.ts:31]: Good point; will add some :+1:

File: .github/workflows/ci.yml (hunk 9 above)
Comment (arieljbirnbaum) [reply .github/workflows/ci.yml:590]: Yes, to make sure that it doesn't go stale with any future changes (so that it's ready to go whenever we need it again). i.e. it's a test _of_ the stress logic rather than a stress test of the logic :)

The test starts three threads, each thread does two iterations; in total it shouldn't be a heavy burden on CI; I think :thinking: 

Then again, what's probably more expensive is just to start the container for the test, so maybe it could be grouped with some other step?

Comment (arieljbirnbaum) [reply .github/workflows/ci.yml:590]: /cc @Verin1005 to address your [comment](https://github.com/litentry/litentry-parachain/pull/2054#pullrequestreview-1594773483); I don't think it should add much more maintenance overhead beyond keeping the _regular_ tests up to date. There is some duplication right now but I'm planning to tackle it with #1878 and #1153

Comment (Kailai-Wang) [reply .github/workflows/ci.yml:590]: It's fine as long as it's not running extremely long 😎

File: tee-worker/ts-tests/stress/test/test.ts (hunk 10 above)
Comment (arieljbirnbaum) [reply tee-worker/ts-tests/stress/test/test.ts:31]: @Verin1005 I added comments throughout the test to clarify the purpose of each section; does that help?

Comment (0xverin) [reply tee-worker/ts-tests/stress/test/test.ts:31]: Thanks!I think it's actually possible, because your function and variable naming is actually kind of a commentary.:p

File: tee-worker/ts-tests/stress-runner (hunk 3 above)
Comment (arieljbirnbaum) [reply tee-worker/ts-tests/stress-runner:3]: dropped for now

**Review by 0xverin — APPROVED** [review]
LGTM.Thanks!
